crypto/s2n_ecc_evp.c

```diff
@@ -0,0 +1,390 @@
+#include <stdint.h>
+
+#include "tls/s2n_kex.h"
+#include "tls/s2n_tls_parameters.h"
+#include "utils/s2n_mem.h"
+#include "utils/s2n_safety.h"
+
+#define TLS_EC_EVP_CURVE_TYPE_NAMED S2N_ECC_EVP_SUPPORTED_CURVES_COUNT
+
+/* IANA values can be found here:
+ * https://tools.ietf.org/html/rfc8446#appendix-B.3.1.4 */
+/* Share sizes are described here:
+ * https://tools.ietf.org/html/rfc8446#section-4.2.8.2 and include the extra
+ * "legacy_form" byte */
+#if S2N_OPENSSL_VERSION_AT_LEAST(1, 1, 0) && !defined(LIBRESSL_VERSION_NUMBER)
+const struct s2n_ecc_named_curve
+    s2n_ecc_evp_supported_curves[S2N_ECC_EVP_SUPPORTED_CURVES_COUNT] = {
+        {.iana_id = TLS_EC_CURVE_SECP_256_R1,
+         .libcrypto_nid = NID_X9_62_prime256v1,
+         .name = "secp256r1",
+         .share_size = (32 * 2) + 1},
+        {.iana_id = TLS_EC_CURVE_SECP_384_R1,
+         .libcrypto_nid = NID_secp384r1,
+         .name = "secp384r1",
+         .share_size = (48 * 2) + 1},
+        {.iana_id = TLS_EC_CURVE_ECDH_X25519,
+         .libcrypto_nid = NID_X25519,
+         .name = "x25519",
+         .share_size = 32},
+};
+#else
+const struct s2n_ecc_named_curve
+    s2n_ecc_evp_supported_curves[S2N_ECC_EVP_SUPPORTED_CURVES_COUNT] = {
+        {.iana_id = TLS_EC_CURVE_SECP_256_R1,
+         .libcrypto_nid = NID_X9_62_prime256v1,
+         .name = "secp256r1",
+         .share_size = (32 * 2) + 1},
+        {.iana_id = TLS_EC_CURVE_SECP_384_R1,
+         .libcrypto_nid = NID_secp384r1,
+         .name = "secp384r1",
+         .share_size = (48 * 2) + 1},
+};
+#endif
+
+static EVP_PKEY *
+s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_curve);
+static int s2n_ecc_evp_compute_shared_secret(EVP_PKEY *own_key,
+                                             EVP_PKEY *peer_public,
+                                             struct s2n_blob *shared_secret);
+
+static EVP_PKEY *
+s2n_ecc_evp_generate_own_key(const struct s2n_ecc_named_curve *named_curve) {
+  EVP_PKEY *evp_pkey = NULL;
+  EVP_PKEY_CTX *pctx = NULL;
+  EVP_PKEY_CTX *kctx = NULL;
+  EVP_PKEY *params = NULL;
+
+#if S2N_OPENSSL_VERSION_AT_LEAST(1, 1, 0) && !defined(LIBRESSL_VERSION_NUMBER)
+  if (named_curve->libcrypto_nid == NID_X25519) {
+    pctx = EVP_PKEY_CTX_new_id(named_curve->libcrypto_nid, NULL);
+    if (pctx == NULL) {
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_keygen_init(pctx) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_keygen(pctx, &evp_pkey) != 1) {
+      EVP_PKEY_CTX_free(pctx);
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+
+    EVP_PKEY_CTX_free(pctx);
+    return evp_pkey;
+  }
+#endif
+  if (named_curve->libcrypto_nid == NID_X9_62_prime256v1 ||
+      named_curve->libcrypto_nid == NID_secp384r1) {
+    pctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    if (pctx == NULL) {
+      S2N_ERROR_PTR(S2N_ERR_ECDHE_GEN_KEY);
+    }
+    if (EVP_PKEY_paramgen_init(pctx) != 1) {
```

Comment:
same for the suggestions as above, we should be able to clean up some lines with DEFER_CLEANUP